[WIP]Orders & Shipments endpoints

sandervanhooft opened the pull request:
**Don't merge yet**, review only please. See [here](https://github.com/mollie/mollie-api-php/issues/247#issuecomment-413201751).

Relates to #247.

To do's:
- [x] add `method` features (field, perhaps some methods) to `Order` resource ([docs](
https://github.com/mollie/api-documentation/commit/919b7021cf33e4b95c4ebc6e515994bbab0ad1bf))
- [x] add `tracking` features (fields, perhaps some methods) to `Shipment` resource
- [x] implement `Order.shipAll()` shorthand
- [x] Write examples (`/examples/...`)
- [x] remove `Order.isCancelable()` (when `isCancelable` field implemented on endpoint). Add isCancelable to `Order` fixture.
- [x] remove `OrderLine.isCancelable()` (when `isCancelable` field implemented on endpoint). Add isCancelable to `OrderLine` fixture.
- [x] Add helper methods `get($lineId)` and `isCancelable($lineId)` to `OrderLineCollection`.
- [x] Check Order and Shipment statuses. I.e. `void` (dropped) and `expired` (introduced).
- [x] Add `refunds` features to `Order`
- [ ] Incorporate suggested changes (code review)
- [ ] Manually test examples when API is live _(pending)_

sandervanhooft commented:
It seems to me the Order.ship shorthand method should be removed, as the shipment endpoint is becoming more advanced ([link](https://github.com/mollie/api-documentation/pull/165))

Smitsel commented:
@sandervanhooft most of the time a whole order will be shipped in it's entirety. So i quite like the shorthand.

Smitsel commented on src/Resources/OrderLine.php:
```diff
@@ -0,0 +1,293 @@
+
+    public function cancel()
+    {
+        return $this->client->performHttpCall(
```
Maybe better to use an endpoint here?
`$this->client->orderLines->cancelFor()`

sandervanhooft replied:
It's possible, but that would require an additional call to retrieve the parent order. That's why I went with the `performHttpCall` approach here, so I could use the `orderID` instead.

Smitsel replied:
Allright!

Smitsel commented on src/Resources/Shipment.php:
```diff
@@ -0,0 +1,72 @@
+     * @var object[]
+     */
+    public $_links;
+
```
Just a reminder that we should add the new Tracking fields :)

sandervanhooft replied:
Done

Smitsel commented on src/Resources/Shipment.php:
```diff
@@ -0,0 +1,72 @@
+     */
+    public function order()
+    {
+        return $this->client->orders->get($this->orderId);
```
Most of the times in the client we use the provided _links url to achieve this.
Ill leave it up to you if you want that to be consistent.

sandervanhooft replied:
I found it far more readable this way :)

willemstuursma commented on src/Resources/Order.php:
```diff
@@ -0,0 +1,298 @@
+<?php
```
@sandervanhooft we've added `method` https://github.com/mollie/api-documentation/commit/919b7021cf33e4b95c4ebc6e515994bbab0ad1bf

sandervanhooft replied:
Done

sandervanhooft commented on src/Endpoints/OrderEndpoint.php:
```diff
@@ -0,0 +1,120 @@
+     * @return null
+     * @throws ApiException
+     */
+    public function delete($orderId)
```
Will drop `Order.Cancel()`.

sandervanhooft replied:
Meant to drop `Order.delete()` obviously.

sandervanhooft replied:
Done

sandervanhooft commented on src/Endpoints/OrderEndpoint.php:
```diff
@@ -0,0 +1,120 @@
+     * Cancel the given Order. This is just an alias of the 'delete' method.
+     *
+     * Will throw a ApiException if the order id is invalid or the resource cannot be found.
+     * Returns with HTTP status No Content (204) if successful.
```
Done

sandervanhooft commented on src/Endpoints/OrderLineEndpoint.php:
```diff
@@ -0,0 +1,52 @@
+    }
+
+    /**
+     * @param Order $order
```
Done

sandervanhooft commented on src/Resources/Order.php:
```diff
@@ -0,0 +1,298 @@
+     *
+     * @return bool
+     */
+    public function isCancelable()
```
That's great! Leave it as is for now?

Smitsel replied:
Yes, ill ping you in the api-documentation PR when its merged.

sandervanhooft commented on src/Endpoints/ShipmentEndpoint.php:
```diff
@@ -0,0 +1,82 @@
+use Mollie\Api\Resources\Shipment;
+use Mollie\Api\Resources\ShipmentCollection;
+
+class ShipmentEndpoint extends EndpointAbstract
```
Done

sandervanhooft commented on src/Resources/Order.php:
```diff
@@ -0,0 +1,298 @@
+     *
+     * @return Shipment
+     */
+    public function createShipment(array $options = [])
```
I think you're talking about the lines parameter in the docs ([here](https://docs.mollie.com/reference/v2/shipments-api/create-shipment)):

> Lines: An array of objects containing the order line details you want to create a shipment for. If you send an empty array, the entire order will be shipped. If the order is already partially shipped, any remaining lines will be shipped.

I'd drop the `ship()` and `shipLines(...)` shortcuts for now and go for the simplest solution for now: the existing `createShipment(...)`.

Perhaps we can add a `shipAll()` or `shipRemaining()` later if there's a need for it.

sandervanhooft replied:
I'll include a comment about what happens when the `lines` variable is an empty array in the `createShipment` method docblock.

sandervanhooft commented on src/Resources/Order.php:
```diff
@@ -0,0 +1,298 @@
+     *
+     * @return Shipment
+     */
+    public function ship(array $lineIds)
```
I'll drop this method for now.

sandervanhooft replied:
Done.

sandervanhooft commented on src/Resources/Order.php:
```diff
@@ -0,0 +1,298 @@
+    }
+
+    /**
+     * @param array $parameters
```
Done

sandervanhooft commented on src/Resources/Order.php:
```diff
@@ -0,0 +1,298 @@
+
+use Mollie\Api\Types\OrderStatus;
+
+class Order extends BaseResource
```
Done

sandervanhooft commented:
To do's:
- [ ] update `Order.isCancelable()` (when `isCancelable` field implemented on endpoint)
- [ ] update `Shipment.isCancelable()` (when `isCancelable` field implemented on endpoint)
- [ ] add `tracking` features (fields, perhaps some methods) to `Shipment` resource
- [ ] add `method` features (field, perhaps some methods) to `Order` resource ([docs](
https://github.com/mollie/api-documentation/commit/919b7021cf33e4b95c4ebc6e515994bbab0ad1bf))
- [ ] implement `Order.shipAll()` shorthand

sandervanhooft commented:
Just received beta API access. Will continue this Monday.

sandervanhooft commented:
Dropped `isCancelable()` method from `Order` and `OrderLine` in favor of `isCancelable` attribute.

Added `get($lineId)` and `isCancelable($lineId)` helpers to `OrderLineCollection`, so you can now check before you cancel a line:

```php
$order = $mollie->orders->get($orderId);
if ($order->lines()->isCancelable($lineId)) { // this is new :)
    $order->cancelLine($lineId);
}
```

sandervanhooft commented:
@ndijkstra @Smitsel 
I've distilled a basic end-to-end test script from the orders and shipment examples.

The Orders client implementation seems fine. I am unable to adequately test the Shipment features.

## Tested (all ok)
- 26-new-order
- 27-handle-order-status-change
- 29-list-orders
- 30-cancel-order-line

## Tested (with problems, see Findings below)
- 28-cancel-order
- 31-ship-order-completely
- 32-ship-order-partially
- 33-get-shipment
- 34-list-order-shipments

## The script

```php
$mollie = new \Mollie\Api\MollieApiClient();
$mollie->setApiKey("test_VWHhavyaJ3rtaSpuzkan7qHjy99ED2");

$order = $mollie->orders->create([
    "amount" => [
      "value" => "1127.99",
      "currency" => "EUR"
    ],
    "billingAddress" => [
      "streetAndNumber" => "Keizersgracht 313",
      "postalCode" => "1016 EE",
      "city" => "Amsterdam",
      "country" => "nl",
      "givenName" => "Luke",
      "familyName" => "Skywalker",
      "email" => "luke@skywalker.com",
    ],
    "shippingAddress" => [
      "streetAndNumber" => "Keizersgracht 313",
      "postalCode" => "1016 EE",
      "city" => "Amsterdam",
      "country" => "nl",
      "givenName" => "Luke",
      "familyName" => "Skywalker",
      "email" => "luke@skywalker.com",
    ],
    "metadata" => [
      "order_id" => "1337",
      "description" => "Lego cars"
    ],
    "consumerDateOfBirth" => "1958-01-31",
    "locale" => "nl_NL",
    "orderNumber" => "1337",
    "redirectUrl" => "https://example.org/redirect",
    "webhookUrl" => "https://example.org/webhook",
    "method" => "ideal",
    "lines" => [
        [
            "sku" => "5702016116977",
            "name" => "LEGO 42083 Bugatti Chiron",
            "productUrl" => "https://shop.lego.com/nl-NL/Bugatti-Chiron-42083",
            "imageUrl" => 'https://sh-s7-live-s.legocdn.com/is/image//LEGO/42083_alt1?$main$',
            "quantity" => 2,
            "vatRate" => "21.00",
            "unitPrice" => [
                "currency" => "EUR",
                "value" => "399.00"
            ],
            "totalAmount" => [
                "currency" => "EUR",
                "value" => "798.00"
            ],
            "vatAmount" => [
                "currency" => "EUR",
                "value" => "138.50"
            ]
        ],
        [
            "type" => "digital",
            "sku" => "5702015594028",
            "name" => "LEGO 42056 Porsche 911 GT3 RS",
            "productUrl" => "https://shop.lego.com/nl-NL/Porsche-911-GT3-RS-42056",
            "imageUrl" => 'https://sh-s7-live-s.legocdn.com/is/image/LEGO/42056?$PDPDefault$',
            "quantity" => 1,
            "vatRate" => "21.00",
            "unitPrice" => [
                "currency" => "EUR",
                "value" => "329.99"
            ],
            "totalAmount" => [
                "currency" => "EUR",
                "value" => "329.99"
            ],
            "vatAmount" => [
                "currency" => "EUR",
                "value" => "57.27"
            ]
        ]
    ]
]);

$order->getCheckoutUrl();

$order->lines()->isCancelable($order->lines()[0]->id);
$order->cancelLine($order->lines()[0]->id);
$order = $mollie->orders->get($order->id);
$order->lines()->isCancelable($order->lines()[0]->id);

$order = $mollie->orders->get($order->id);
$order->isCancelable;
$order = $mollie->orders->cancel($order->id);

$order = $mollie->orders->get($order->id);
$order->isCancelable;

$mollie->orders->page();
$mollie->orders->page()->next();

$partial_shipment = $order->createShipment(
    [
        'lines' => [
            [
                'id' => $order->lines()[0]->id,
                'quantity' => 1,
            ],
        ],
    ]
);

$complete_shipment = $order->shipAll();

$order->getShipment($partial_shipment);
$order->getShipment($complete_shipment);
$order->shipments();
```

## Findings

```php
$order->getCheckoutUrl(); // works, but actually visiting the url does not, so I cannot update the status / perform shipments (requires status paid/authorized)

$order->cancel(); // method is missing on resource, use $mollie->orders->cancel($id) instead. Nice to have though.

$order->refresh(); // would be nice (via _links->self ? on baseresource?)

$order = $mollie->orders->cancel($order->id); // 500 internal server error

$partial_shipment = $order->createShipment(
    [
        'lines' => [
            [
                'id' => $order->lines()[0]->id,
                'quantity' => 1,
            ],
        ],
    ]
); // Results in Mollie/Api/Exceptions/ApiException with message 'Error executing API call (400: Bad Request): Order line shipment 0 contains invalid data: Order line odl_od4qs1 cannot be shipped from state: created. Field: lines.0.id. Documentation: https://docs.mollie.com/guides/handling-errors'
```

sandervanhooft commented:
Just stumbled upon the `refunds` features in the API reference. Added it to the to do list.

Smitsel commented on examples/orders/.gitignore:
```diff
@@ -1 +0,0 @@
-*.txt
```
This is where the .txt database is kept. We want these ignored right?

sandervanhooft replied:
Didn't know about that database. Reinstating the `orders` folder with `.gitignore`.
Done

Smitsel commented on src/Endpoints/OrderEndpoint.php:
```diff
@@ -0,0 +1,108 @@
+    }
+
+    /**
+     * Cancel the given Order.
```
We can still cancel a partially shipped order.
The order status will then result in `completed`.

Can we maybe describe this in this doc block?

sandervanhooft replied:
Done!

Smitsel commented on src/Resources/Order.php:
```diff
@@ -0,0 +1,388 @@
+     * @param  string $lineId
+     * @return null
+     */
+    public function cancelLine($lineId)
```
Should we add a quantity here now that we support partial cancels?

sandervanhooft replied:
Hi @Smitsel 

I can't find anything on partial order line cancellation with an explicitly _defined_ quantity parameter in the docs ([here](https://docs.mollie.com/reference/v2/orders-api/cancel-order-line)).

Is this not documented (yet)?

Or are you referring to first partially shipping and then _canceling the remainder of the order_? This would not require a quantity parameter.

Smitsel replied:
Sorry, i forgot to update the documentation.
Here it is: https://github.com/mollie/api-documentation/pull/192

sandervanhooft replied:
I've added a `$data = []` for consistency / future proofing.
+ examples and tests for canceling an order line partially.

Smitsel commented on src/Resources/OrderLine.php:
```diff
@@ -0,0 +1,284 @@
+        return $this->type === OrderLineType::TYPE_SURCHARGE;
+    }
+
+    public function cancel()
```
We can give a quantity to this endpoint to partially cancel a line.

sandervanhooft replied:
Similar to my other comment above: this is not documented yet. Are you sure? / What does the `curl` call look like?

Smitsel replied:
See other comment ;)

sandervanhooft replied:
I've added a `$data = []` for consistency / future proofing.
+ examples and tests for canceling an order line partially.

Smitsel commented on src/Resources/OrderLineCollection.php:
```diff
@@ -0,0 +1,47 @@
+     * @param  string $lineId
+     * @return OrderLine|null
+     */
+    public function get($lineId)
```
I can see where this `get` on a Collection is useful. Maybe put it in the abstraction?

sandervanhooft replied:
I agree this method is helpful on collections that are always returned completely.

For paginated collections it may be a bit confusing. `get` may then return null if the `id` is not in the collection, while the resource may exist at the server.

We could put the method in a trait, or implement a more advanced one in the `CursorCollection` abstraction. Or just leave it as it is and see if we're going to need it.

Bear in mind that it is implemented here because it is not possible to directly `GET` the order line resource via a dedicated endpoint (i.e. `GET /orders/{id}/lines/{id}`). Imho it's a bit of an odd case 😄 .

sandervanhooft replied:
@Smitsel Let me know what direction to take here.

Smitsel replied:
Just leave it like this and we'll see ;)

sandervanhooft replied:
Ok!

Smitsel commented on src/Resources/OrderLineCollection.php:
```diff
@@ -0,0 +1,47 @@
+     * @param  string  $lineId
+     * @return boolean
+     */
+    public function isCancelable($lineId)
```
Don't really see the use for this. I would just use the isCancelable on the Orderline value object.

Smitsel replied:
I see now that this is used for something like: `$orders->lines()->isCancelable("odl_asdfa");`
But i think this is a bit vague. So i'd rather see `$line = $orders->lines()->get("odl_xxx");`.
And then `if ($line->isCancelable)`

sandervanhooft replied:
It seemed to make sense in the example [here](https://github.com/mollie/mollie-api-php/blob/196fc17e05ffd15bd469a6a53ae4700b72a3ec79/examples/30-cancel-order-line.php#L22), so I added it. But it may be too much of a personal preference.

It would take an extra line with a separate `get(..)` call to do the same.

I'm fine with removing it, just let me know.

sandervanhooft replied:
Done.

Smitsel commented:
Everything looks really good, just some minor stuff.
Only major thing is that the partial cancelations are missing.

sandervanhooft commented:
Thanks @Smitsel !

I'll be occupied at Laracon.eu the coming days - in a shiny Mollie shirt so it seems 😎 .

Will look at this PR at the beginning of next week. Contact me if this is causing any problems with the release planning.
